tests/test-integration/src/test/java/org/eclipse/jetty/test/MultiAuthenticatorTest.java

```diff
@@ -0,0 +1,306 @@
+        _client.start();
+    }
+
+    @AfterEach
+    public void after() throws Exception
+    {
+        _client.stop();
+        _server.stop();
+    }
+
+    @Test
+    public void testMultiAuthentication() throws Exception
+    {
+        URI uri = URI.create("http://localhost:" + _connector.getLocalPort());
+        ContentResponse response = _client.GET(uri);
+        assertThat(response.getStatus(), equalTo(HttpStatus.OK_200));
+        assertThat(response.getContentAsString(), containsString("<h1>Multi Login Page</h1>"));
+        assertThat(response.getContentAsString(), containsString("/login/openid"));
+        assertThat(response.getContentAsString(), containsString("/login/form"));
+
+        // Try Form Login.
+        response = _client.GET(uri.resolve("/login/form"));
+        assertThat(response.getStatus(), equalTo(HttpStatus.OK_200));
+        assertThat(response.getContentAsString(), containsString("<form action=\"j_security_check\" method=\"POST\">"));
+
+        // Form login is successful.
+        Fields fields = new Fields();
+        fields.put("j_username", "user");
+        fields.put("j_password", "password");
+        response = _client.POST(uri.resolve("/j_security_check"))
+            .body(new FormRequestContent(fields))
+            .send();
+        assertThat(response.getStatus(), equalTo(HttpStatus.OK_200));
+        assertThat(response.getContentAsString(), containsString("userPrincipal: user"));
+        assertThat(response.getContentAsString(), containsString("MultiAuthenticator$MultiSucceededAuthenticationState"));
+
+        // Logout is successful.
+        response = _client.GET(uri.resolve("/logout"));
+        assertThat(response.getStatus(), equalTo(HttpStatus.OK_200));
+        assertThat(response.getContentAsString(), containsString("<h1>Multi Login Page</h1>"));
+        assertThat(response.getContentAsString(), containsString("/login/openid"));
+        assertThat(response.getContentAsString(), containsString("/login/form"));
+
+        // We can now log in with OpenID.
+        _provider.setUser(new OpenIdProvider.User("UserId1234", "openIdUser"));
+        response = _client.GET(uri.resolve("/login/openid"));
+        assertThat(response.getStatus(), equalTo(HttpStatus.OK_200));
+        assertThat(response.getContentAsString(), containsString("userPrincipal: UserId1234"));
+        assertThat(response.getContentAsString(), containsString("Authenticated with OpenID"));
+        assertThat(response.getContentAsString(), containsString("name: openIdUser"));
+
+        // Logout is successful.
+        response = _client.GET(uri.resolve("/logout"));
+        assertThat(response.getStatus(), equalTo(HttpStatus.OK_200));
+        assertThat(response.getContentAsString(), containsString("<h1>Multi Login Page</h1>"));
+        assertThat(response.getContentAsString(), containsString("/login/openid"));
+        assertThat(response.getContentAsString(), containsString("/login/form"));
+    }
```

Comment:
I would like to see:

* for this test, the access to a protected resource should be tried for both authentications.
* a new test for failed authentication, protected resources are not accessible
* a test where one authentication succeeds but the other fails -- can I still access the resource? Basically I would like to know if "multi" has "and" semantic (all authentications but must successful), or "or" semantic (one successful authentication is enough).